File: python/tests/queue_test.py
"""
Tests for the queue class.

https://bbc.github.io/cloudfit-public-docs/asyncio/testing.html
"""

from asyncio import Future
import redis.asyncio as redis
from bullmq import Queue, Worker, Job
from uuid import uuid4

import asyncio
import unittest
import os
import time

queueName = ""
prefix = os.environ.get('BULLMQ_TEST_PREFIX') or "bull"

class TestQueue(unittest.IsolatedAsyncioTestCase):

    def setUp(self):
        print("Setting up test queue")
        queueName = f"__test_queue__{uuid4().hex}"

    async def asyncTearDown(self):
        connection = redis.Redis(host='localhost')
        await connection.flushdb()

    async def test_connection_str(self):
        queue = Queue(queueName, {"connection": "redis://localhost:6379", "prefix": prefix})

        self.assertIsInstance(queue, Queue)
        await queue.close()

    async def test_add_job(self):
        queue = Queue(queueName, {"prefix": prefix})
        job = await queue.add("test-job", {"foo": "bar"}, {})

        self.assertEqual(job.id, "1")
        await queue.close()

    async def test_get_jobs(self):
        queue = Queue(queueName, {"prefix": prefix})
        job1 = await queue.add("test-job", {"foo": "bar"}, {})
        job2 = await queue.add("test-job", {"foo": "bar"}, {})
        jobs = await queue.getJobs(["wait"])

        self.assertEqual(job2.id, jobs[0].id)
        self.assertEqual(job1.id, jobs[1].id)
        await queue.close()

    async def test_get_job_state(self):
        queue = Queue(queueName, {"prefix": prefix})
        job = await queue.add("test-job", {"foo": "bar"}, {})
        state = await queue.getJobState(job.id)

        self.assertEqual(state, "waiting")
        await queue.close()

    async def test_add_job_with_options(self):
        queue = Queue(queueName, {"prefix": prefix})
        data = {"foo": "bar"}
        attempts = 3
        delay = 1000
        job = await queue.add("test-job", data=data, opts={"attempts": attempts, "delay": delay})

        self.assertEqual(job.id, "1")
        self.assertEqual(job.attempts, attempts)
        self.assertEqual(job.delay, delay)
        self.assertEqual(job.data, data)

        await queue.close()

    async def test_is_paused(self):
        queue = Queue(queueName, {"prefix": prefix})
        await queue.pause()
        isPaused = await queue.isPaused()

        self.assertEqual(isPaused, True)

        await queue.resume()

        isPaused = await queue.isPaused()

        self.assertEqual(isPaused, False)

        await queue.close()

    async def test_is_paused_with_custom_prefix(self):
        custom_prefix = "{" + prefix + "}"
        queue = Queue(queueName, {"prefix": custom_prefix})
        await queue.pause()
        isPaused = await queue.isPaused()

        self.assertEqual(isPaused, True)

        await queue.resume()

        isPaused = await queue.isPaused()

        self.assertEqual(isPaused, False)

        await queue.obliterate()
        await queue.close()

    async def test_trim_events_manually(self):
        queue = Queue(queueName, {"prefix": prefix})
        await queue.add("test", data={}, opts={})
        await queue.add("test", data={}, opts={})
        await queue.add("test", data={}, opts={})
        await queue.add("test", data={}, opts={})

        events_length = await queue.client.xlen(f"{queue.prefix}:{queueName}:events")
        self.assertEqual(events_length, 8)

        await queue.trimEvents(0)

        events_length = await queue.client.xlen(f"{queue.prefix}:{queue.name}:events")

        self.assertEqual(events_length, 0)

        await queue.close()

    async def test_trim_events_manually_with_custom_prefix(self):
        custom_prefix = "{" + prefix + "}"
        queue = Queue(queueName, {"prefix": custom_prefix})
        await queue.add("test", data={}, opts={})
        await queue.add("test", data={}, opts={})
        await queue.add("test", data={}, opts={})
        await queue.add("test", data={}, opts={})

        events_length = await queue.client.xlen(f"{custom_prefix}:{queueName}:events")
        self.assertEqual(events_length, 8)

        await queue.trimEvents(0)

        events_length = await queue.client.xlen(f"{custom_prefix}:{queue.name}:events")

        self.assertEqual(events_length, 0)

        await queue.obliterate()
        await queue.close()

    async def test_get_delayed_count(self):
        queue = Queue(queueName, {"prefix": prefix})
        data = {"foo": "bar"}
        delay = 1000
        await queue.add("test-job", data=data, opts={"delay": delay})
        await queue.add("test-job", data=data, opts={"delay": delay * 2})

        count = await queue.getDelayedCount()
        self.assertEqual(count, 2)

        await queue.close()

    async def test_retry_failed_jobs(self):
        queue = Queue(queueName, {"prefix": prefix})
        job_count = 8

        fail = True

        async def process(job: Job, token: str):
            await asyncio.sleep(1)
            if fail:
                raise Exception("failed")
            return
        order = 0

        worker = Worker(queueName, process, {"prefix": prefix})

        failed_events = Future()

        def failing(job: Job, result):
            nonlocal order
            if order == (job_count - 1):
                failed_events.set_result(None)
            order += 1

        worker.on("failed", failing)

        for index in range(job_count):
            data = {"idx": index}
            await queue.add("test", data=data)

        await failed_events

        worker.off('failed', failing)

        failed_count = await queue.getFailedCount()

        self.assertEqual(failed_count, job_count)

        order = 0

        completed_events = Future()

        def completing(job: Job, result):
            nonlocal order
            if order == (job_count - 1):
                completed_events.set_result(None)
            order += 1

        worker.on("completed", completing)

        fail = False

        await queue.retryJobs({'count': 2})

        await completed_events

        worker.off('completed', completing)

        completed_count = await queue.getJobCounts('completed')
        self.assertEqual(completed_count['completed'], job_count)

        await queue.close()
        await worker.close()

    async def test_retry_completed_jobs(self):
        queue = Queue(queueName, {"prefix": prefix})
        job_count = 8

        async def process(job: Job, token: str):
            await asyncio.sleep(1)
            return
        order = 0

        worker = Worker(queueName, process, {"prefix": prefix})

        completed_events1 = Future()

        def completing1(job: Job, result):
            nonlocal order
            if order == (job_count - 1):
                completed_events1.set_result(None)
            order += 1

        worker.on("completed", completing1)

        for index in range(job_count):
            data = {"idx": index}
            await queue.add("test", data=data)

        await completed_events1

        worker.off('completed', completing1)

        completed_count1 = await queue.getJobCounts('completed')
        self.assertEqual(completed_count1['completed'], job_count)

        order = 0

        completed_events2 = Future()

        def completing2(job: Job, result):
            nonlocal order
            if order == (job_count - 1):
                completed_events2.set_result(None)
            order += 1

        worker.on("completed", completing2)

        await queue.retryJobs({'count': 2, 'state': 'completed'})

        await completed_events2

        worker.off('completed', completing2)

        completed_count2 = await queue.getJobCounts('completed')
        self.assertEqual(completed_count2['completed'], job_count)

        await queue.close()
        await worker.close()

    async def test_retry_failed_jobs_before_timestamp(self):
        queue = Queue(queueName, {"prefix": prefix})
        job_count = 8

        fail = True

        async def process(job: Job, token: str):
            await asyncio.sleep(1)
            if fail:
                raise Exception("failed")
            return
        order = 0

        worker = Worker(queueName, process, {"prefix": prefix})

        failed_events = Future()
        timestamp = 0

        def failing(job: Job, result):
            nonlocal order
            nonlocal timestamp
            if order == (job_count - 1):
                failed_events.set_result(None)
            if job.data['idx'] == (job_count/2 - 1):
                timestamp = round(time.time() * 1000)
            order += 1

        worker.on("failed", failing)

        for index in range(job_count):
            data = {"idx": index}
            await queue.add("test", data=data)

        await failed_events

        worker.off('failed', failing)

        failed_count = await queue.getJobCounts('failed')

        self.assertEqual(failed_count['failed'], 8)

        order = 0

        completed_events = Future()

        def completing(job: Job, result):
            nonlocal order
            if order == (job_count/2 - 1):
                completed_events.set_result(None)
            order += 1

        worker.on("completed", completing)

        fail = False

        await queue.retryJobs({'count': 2, 'timestamp': timestamp})

        await completed_events

        worker.off('completed', completing)

        completed_count = await queue.getCompletedCount()
        self.assertEqual(completed_count, 4)

        await queue.close()
        await worker.close()

    async def test_retry_jobs_when_queue_is_paused(self):
        queue = Queue(queueName, {"prefix": prefix})
        job_count = 8

        fail = True

        async def process(job: Job, token: str):
            await asyncio.sleep(1)
            if fail:
                raise Exception("failed")
            return
        order = 0

        worker = Worker(queueName, process, {"prefix": prefix})

        failed_events = Future()

        def failing(job: Job, result):
            nonlocal order
            if order == (job_count - 1):
                failed_events.set_result(None)
            order += 1

        worker.on("failed", failing)

        for index in range(job_count):
            data = {"idx": index}
            await queue.add("test", data=data)

        await failed_events

        worker.off('failed', failing)

        failed_count = await queue.getJobCounts('failed')

        self.assertEqual(failed_count['failed'], 8)

        order = 0

        fail = False

        await queue.pause()
        await queue.retryJobs({'count': 2})

        paused_count = await queue.getJobCounts('paused')
        self.assertEqual(paused_count['paused'], job_count)

        await queue.close()
        await worker.close()

    async def test_promote_all_delayed_jobs(self):
        queue = Queue(queueName, {"prefix": prefix})
        job_count = 8

        for index in range(job_count):
            data = { "idx": index }
            await queue.add("test", data=data, opts={ "delay": 5000 })

        delayed_count = await queue.getJobCounts('delayed')
        self.assertEqual(delayed_count['delayed'], job_count)

        await queue.promoteJobs()

        waiting_count = await queue.getJobCounts('waiting')
        self.assertEqual(waiting_count['waiting'], job_count)

        async def process(job: Job, token: str):
            await asyncio.sleep(0.1)
            return
        order = 0

        worker = Worker(queueName, process, {"prefix": prefix})

        completed_events = Future()

        def completing(job: Job, result):
            nonlocal order
            if order == (job_count - 1):
                completed_events.set_result(None)
            order += 1

        worker.on("completed", completing)

        await completed_events

        worker.off('completed', completing)

        delayed_count = await queue.getJobCounts('delayed')

        self.assertEqual(delayed_count['delayed'], 0)

        await queue.close()
        await worker.close()

    async def test_remove_job(self):
        queue = Queue(queueName, {"prefix": prefix})
        job = await queue.add("test", {"foo": "bar"}, {})
        await queue.remove(job.id)
        job = await Job.fromId(queue, job.id)
        self.assertIsNone(job)

        waiting_count = await queue.getWaitingCount()

        self.assertEqual(waiting_count, 0)

        await queue.close()

    async def test_get_counts_per_priority(self):
        queue = Queue(queueName, {"prefix": prefix})
        jobs = [{
            "name": "test",
            "data": {},
            "opts": {
                "priority": index % 4
            }
        } for index in range(42)]
        await queue.addBulk(jobs)
        counts = await queue.getCountsPerPriority([0, 1, 2, 3])
        self.assertEqual(counts, {
            "0": 11,
            "1": 11,
            "2": 10,
            "3": 10
        })

        await queue.close()

    async def test_reusable_redis(self):
        conn = redis.Redis(decode_responses=True, host="localhost", port="6379", db=0)
        queue = Queue(queueName, {"connection": conn, "prefix": prefix})
        job = await queue.add("test-job", {"foo": "bar"}, {})

        self.assertEqual(job.id, "1")
        await queue.close()

    async def test_obliterate_with_force_false_should_fail_with_active_jobs(self):
        """Test that obliterate without force fails when there are active jobs"""
        queue = Queue(queueName, {"prefix": prefix})
        
        # Add jobs
        await queue.add("test-job", {"foo": "bar"}, {})
        await queue.add("test-job", {"qux": "baz"}, {})
        
        # Create a worker that processes jobs slowly
        async def process(job: Job, token: str):
            await asyncio.sleep(1)
            return "done"
        
        worker = Worker(queueName, process, {"prefix": prefix, "autorun": False})

        active_event = Future()
        worker.on("active", lambda job, result: active_event.set_result(None))

        asyncio.ensure_future(worker.run())

        await active_event
        
        # Try to obliterate without force - should fail
        with self.assertRaises(Exception) as context:
            await queue.obliterate(force=False)
        
        self.assertIn("Cannot obliterate queue with active jobs", str(context.exception))
        
        await worker.close()
        await queue.close()

    async def test_obliterate_with_force_true_should_succeed_with_active_jobs(self):
        """Test that obliterate with force=True succeeds even with active jobs"""
        queue = Queue(queueName, {"prefix": prefix})
        
        # Add jobs
        await queue.add("test-job", {"foo": "bar"}, {})
        await queue.add("test-job", {"qux": "baz"}, {})
        await queue.add("test-job", {"foo": "bar2"}, {})
        
        # Create a worker that processes jobs slowly
        async def process(job: Job, token: str):
            await asyncio.sleep(2)
            return "done"
        
        worker = Worker(queueName, process, {"prefix": prefix, "autorun": False})

        active_event = Future()
        worker.on("active", lambda job, result: active_event.set_result(None))

        asyncio.ensure_future(worker.run())

        await active_event
        
        # Verify there are active jobs
        job_counts = await queue.getJobCounts('active')
        self.assertGreater(job_counts.get('active', 0), 0)
        
        # Obliterate with force=True - should succeed
        await queue.obliterate(force=True)
        
        # Verify all keys are deleted
        keys = await queue.client.keys(f"{prefix}:{queueName}:*")
        self.assertEqual(len(keys), 0)
        
        await worker.close()
        await queue.close()

    async def test_obliterate_with_force_true_parameter_conversion(self):
        """Test that force=True is properly converted to string for Redis"""
        queue = Queue(queueName, {"prefix": prefix})
        
        # Add a few jobs
        await queue.add("test-job", {"foo": "bar"}, {})
        await queue.add("test-job", {"foo": "baz"}, {})
        
        await queue.obliterate(force=True)
        
        await queue.close()

    async def test_drain_count_added_unprocessed_jobs(self):
        """Test drain removes all waiting jobs but leaves infrastructure keys"""
        queue = Queue(queueName, {"prefix": prefix})
        max_jobs = 100
        
        # Add jobs with priorities
        for i in range(1, max_jobs + 1):
            await queue.add("test", {"foo": "bar", "num": i}, {"priority": i})
        
        # Check initial counts
        initial_count = await queue.getJobCountByTypes("waiting", "prioritized")
        self.assertEqual(initial_count, max_jobs)
        
        prioritized_count = await queue.getJobCounts("prioritized")
        self.assertEqual(prioritized_count["prioritized"], max_jobs)
        
        # Drain the queue
        await queue.drain()
        
        # Check that all jobs are removed
        count_after_drain = await queue.getJobCountByTypes("waiting", "prioritized")
        self.assertEqual(count_after_drain, 0)
        
        # Check that infrastructure keys remain
        keys = await queue.client.keys(f"{prefix}:{queueName}:*")
        self.assertEqual(len(keys), 5)
        
        # Verify expected infrastructure keys exist
        key_types = []
        for key in keys:
            key_type = key.split(':')[2]
            key_types.append(key_type)
        
        expected_keys = ['marker', 'events', 'meta', 'pc', 'id']
        for expected_key in expected_keys:
            self.assertIn(expected_key, key_types)
        
        await queue.close()

    async def test_drain_delayed_false(self):
        """Test drain with delayed=False keeps delayed jobs"""
        queue = Queue(queueName, {"prefix": prefix})
        max_jobs = 50
        max_delayed_jobs = 50
        
        # Add regular jobs
        for i in range(1, max_jobs + 1):
            await queue.add("test", {"foo": "bar", "num": i}, {})
        
        # Add delayed jobs
        for i in range(1, max_delayed_jobs + 1):
            await queue.add("test", {"foo": "bar", "num": i}, {"delay": 10000})
        
        # Check initial count
        initial_count = await queue.getJobCountByTypes("waiting", "delayed")
        self.assertEqual(initial_count, max_jobs + max_delayed_jobs)
        
        # Drain without delayed jobs
        await queue.drain(delayed=False)
        
        # Check that only delayed jobs remain
        count_after_drain = await queue.getJobCountByTypes("waiting", "delayed")
        self.assertEqual(count_after_drain, max_delayed_jobs)
        
        await queue.close()

    async def test_drain_delayed_true(self):
        """Test drain with delayed=True removes all jobs including delayed"""
        queue = Queue(queueName, {"prefix": prefix})
        max_jobs = 50
        max_delayed_jobs = 50
        
        # Add regular jobs
        for i in range(1, max_jobs + 1):
            await queue.add("test", {"foo": "bar", "num": i}, {})
        
        # Add delayed jobs
        for i in range(1, max_delayed_jobs + 1):
            await queue.add("test", {"foo": "bar", "num": i}, {"delay": 10000})
        
        # Check initial count
        initial_count = await queue.getJobCountByTypes("waiting", "delayed")
        self.assertEqual(initial_count, max_jobs + max_delayed_jobs)
        
        # Drain including delayed jobs
        await queue.drain(delayed=True)
        
        # Check that all jobs are removed
        count_after_drain = await queue.getJobCountByTypes("waiting", "delayed")
        self.assertEqual(count_after_drain, 0)
        
        await queue.close()

    async def test_drain_paused_queue(self):
        """Test drain removes paused jobs when queue is paused"""
        queue = Queue(queueName, {"prefix": prefix})
        max_jobs = 50
        
        # Pause the queue first
        await queue.pause()
        
        # Add jobs (they will go to paused state)
        for i in range(1, max_jobs + 1):
            await queue.add("test", {"foo": "bar", "num": i}, {})
        
        # Check initial count
        initial_count = await queue.getJobCountByTypes("paused")
        self.assertEqual(initial_count, max_jobs)
        
        paused_counts = await queue.getJobCounts("paused")
        self.assertEqual(paused_counts["paused"], max_jobs)
        
        # Drain the queue
        await queue.drain()
        
        # Check that all jobs are removed
        count_after_drain = await queue.getJobCountByTypes("paused")
        self.assertEqual(count_after_drain, 0)
        
        await queue.close()

    async def test_default_job_options(self):
        """Test that defaultJobOptions are applied to jobs added to the queue"""
        queue_name = f"__test_queue__{uuid4().hex}"
        default_attempts = 5
        default_delay = 2000
        queue = Queue(queue_name, {
            "prefix": prefix,
            "defaultJobOptions": {
                "attempts": default_attempts,
                "delay": default_delay
            }
        })
        
        # Add a job without specifying options
        job1 = await queue.add("test-job", {"foo": "bar"})
        
        # Verify that default options were applied
        self.assertEqual(job1.attempts, default_attempts)
        self.assertEqual(job1.delay, default_delay)
        
        # Add a job with custom options that should override defaults
        custom_attempts = 10
        job2 = await queue.add("test-job", {"foo": "baz"}, {"attempts": custom_attempts})
        
        # Verify that custom options override defaults
        self.assertEqual(job2.attempts, custom_attempts)
        self.assertEqual(job2.delay, default_delay)  # Should still use default delay
        
        await queue.close()

if __name__ == '__main__':
    unittest.main()

File: php/tests/QueueTest.php
<?php

declare(strict_types=1);

namespace BullMQ\Tests;

use BullMQ\Queue;
use BullMQ\Job;
use BullMQ\RedisConnection;
use PHPUnit\Framework\TestCase;

/**
 * Integration tests for Queue class.
 * Requires a running Redis server.
 */
class QueueTest extends TestCase
{
    private ?Queue $queue = null;
    private string $queueName;

    protected function setUp(): void
    {
        $this->queueName = 'test-queue-' . uniqid();
        
        try {
            $this->queue = new Queue($this->queueName, [
                'connection' => [
                    'host' => getenv('REDIS_HOST') ?: 'localhost',
                    'port' => (int)(getenv('REDIS_PORT') ?: 6379),
                ],
            ]);
            // Test connection by pinging
            $this->queue->getConnection()->getClient()->ping();
        } catch (\Exception $e) {
            $this->markTestSkipped('Redis server not available: ' . $e->getMessage());
        }
    }

    protected function tearDown(): void
    {
        if ($this->queue !== null) {
            try {
                $this->queue->obliterate(['force' => true]);
            } catch (\Exception $e) {
                // Ignore cleanup errors
            }
            $this->queue->close();
        }
    }

    public function testCanCreateQueue(): void
    {
        $this->assertInstanceOf(Queue::class, $this->queue);
        $this->assertEquals($this->queueName, $this->queue->name);
    }

    public function testCanAddJob(): void
    {
        $job = $this->queue->add('test-job', ['foo' => 'bar']);

        $this->assertInstanceOf(Job::class, $job);
        $this->assertNotNull($job->id);
        $this->assertEquals('test-job', $job->name);
        $this->assertEquals(['foo' => 'bar'], $job->data);
    }

    public function testCanAddJobWithCustomId(): void
    {
        $job = $this->queue->add('test-job', ['foo' => 'bar'], [
            'jobId' => 'custom-id-123',
        ]);

        $this->assertEquals('custom-id-123', $job->id);
    }

    public function testCanAddDelayedJob(): void
    {
        $job = $this->queue->add('delayed-job', ['foo' => 'bar'], [
            'delay' => 60000, // 60 seconds
        ]);

        $this->assertInstanceOf(Job::class, $job);
        $state = $this->queue->getJobState($job->id);
        $this->assertEquals('delayed', $state);
    }

    public function testCanAddPrioritizedJob(): void
    {
        $job = $this->queue->add('priority-job', ['foo' => 'bar'], [
            'priority' => 1,
        ]);

        $this->assertInstanceOf(Job::class, $job);
        $state = $this->queue->getJobState($job->id);
        $this->assertContains($state, ['waiting', 'prioritized']);
    }

    public function testCanGetJob(): void
    {
        $job = $this->queue->add('test-job', ['foo' => 'bar']);
        
        $retrieved = $this->queue->getJob($job->id);

        $this->assertNotNull($retrieved);
        $this->assertEquals($job->id, $retrieved->id);
        $this->assertEquals($job->name, $retrieved->name);
        $this->assertEquals($job->data, $retrieved->data);
    }

    public function testGetJobReturnsNullForNonexistent(): void
    {
        $retrieved = $this->queue->getJob('nonexistent-id');

        $this->assertNull($retrieved);
    }

    public function testCanGetJobCounts(): void
    {
        // Add some jobs
        $this->queue->add('job-1', ['data' => 1]);
        $this->queue->add('job-2', ['data' => 2]);
        $this->queue->add('job-3', ['data' => 3], ['delay' => 60000]);

        $counts = $this->queue->getJobCounts();

        $this->assertArrayHasKey('waiting', $counts);
        $this->assertArrayHasKey('delayed', $counts);
        $this->assertGreaterThanOrEqual(2, $counts['waiting']);
        $this->assertGreaterThanOrEqual(1, $counts['delayed']);
    }

    public function testCanGetJobCountByTypes(): void
    {
        $this->queue->add('job-1', ['data' => 1]);
        $this->queue->add('job-2', ['data' => 2]);

        $count = $this->queue->getJobCountByTypes('waiting');

        $this->assertGreaterThanOrEqual(2, $count);
    }

    public function testCanPauseAndResumeQueue(): void
    {
        $this->assertFalse($this->queue->isPaused());

        $this->queue->pause();
        $this->assertTrue($this->queue->isPaused());

        $this->queue->resume();
        $this->assertFalse($this->queue->isPaused());
    }

    public function testCanRemoveJob(): void
    {
        $job = $this->queue->add('test-job', ['foo' => 'bar']);
        
        $removed = $this->queue->remove($job->id);

        $this->assertTrue($removed);
        $this->assertNull($this->queue->getJob($job->id));
    }

    public function testCanAddBulkJobs(): void
    {
        $jobs = $this->queue->addBulk([
            ['name' => 'bulk-job-1', 'data' => ['id' => 1]],
            ['name' => 'bulk-job-2', 'data' => ['id' => 2]],
            ['name' => 'bulk-job-3', 'data' => ['id' => 3]],
        ]);

        $this->assertCount(3, $jobs);
        foreach ($jobs as $job) {
            $this->assertInstanceOf(Job::class, $job);
            $this->assertNotNull($job->id);
        }
    }

    public function testCanGetWaitingJobs(): void
    {
        $this->queue->add('job-1', ['data' => 1]);
        $this->queue->add('job-2', ['data' => 2]);

        $waitingJobs = $this->queue->getWaiting();

        $this->assertGreaterThanOrEqual(2, count($waitingJobs));
        foreach ($waitingJobs as $job) {
            $this->assertInstanceOf(Job::class, $job);
        }
    }

    public function testCanGetDelayedJobs(): void
    {
        $this->queue->add('delayed-1', ['data' => 1], ['delay' => 60000]);
        $this->queue->add('delayed-2', ['data' => 2], ['delay' => 120000]);

        $delayedJobs = $this->queue->getDelayed();

        $this->assertGreaterThanOrEqual(2, count($delayedJobs));
    }

    public function testCanGetPrioritizedJobs(): void
    {
        $this->queue->add('priority-1', ['data' => 1], ['priority' => 1]);
        $this->queue->add('priority-2', ['data' => 2], ['priority' => 2]);
        $this->queue->add('priority-3', ['data' => 3], ['priority' => 3]);

        $prioritizedJobs = $this->queue->getPrioritized();

        $this->assertGreaterThanOrEqual(3, count($prioritizedJobs));
        foreach ($prioritizedJobs as $job) {
            $this->assertInstanceOf(Job::class, $job);
        }
    }

    public function testCanCleanJobs(): void
    {
        // This test would require jobs to be completed/failed first
        // which requires a worker. For now, just verify the method exists.
        $cleaned = $this->queue->clean(0, 0, 'completed');
        
        $this->assertIsArray($cleaned);
    }

    public function testCanGetCountsPerPriority(): void
    {
        // Add jobs with different priorities
        for ($i = 0; $i < 12; $i++) {
            $this->queue->add('priority-job', ['idx' => $i], [
                'priority' => $i % 4,
            ]);
        }

        $counts = $this->queue->getCountsPerPriority([0, 1, 2, 3]);

        $this->assertArrayHasKey(0, $counts);
        $this->assertArrayHasKey(1, $counts);
        $this->assertArrayHasKey(2, $counts);
        $this->assertArrayHasKey(3, $counts);
        $this->assertEquals(3, $counts[0]);
        $this->assertEquals(3, $counts[1]);
        $this->assertEquals(3, $counts[2]);
        $this->assertEquals(3, $counts[3]);
    }

    public function testCanPromoteDelayedJobs(): void
    {
        // Add delayed jobs
        $this->queue->add('delayed-1', ['data' => 1], ['delay' => 60000]);
        $this->queue->add('delayed-2', ['data' => 2], ['delay' => 60000]);
        $this->queue->add('delayed-3', ['data' => 3], ['delay' => 60000]);

        $counts = $this->queue->getJobCounts();
        $this->assertEquals(3, $counts['delayed']);

        // Promote all delayed jobs
        $this->queue->promoteJobs();

        // Check they moved to waiting
        $counts = $this->queue->getJobCounts();
        $this->assertEquals(0, $counts['delayed']);
        $this->assertGreaterThanOrEqual(3, $counts['waiting']);
    }

    public function testCanObliterateQueue(): void
    {
        // Add some jobs
        $this->queue->add('job-1', ['data' => 1]);
        $this->queue->add('job-2', ['data' => 2]);
        $this->queue->add('job-3', ['data' => 3], ['delay' => 60000]);

        // Verify jobs exist
        $counts = $this->queue->getJobCounts();
        $this->assertGreaterThan(0, $counts['waiting'] + $counts['delayed']);

        // Obliterate the queue
        $this->queue->obliterate(['force' => true]);

        // Verify queue is empty
        $counts = $this->queue->getJobCounts();
        $this->assertEquals(0, $counts['waiting']);
        $this->assertEquals(0, $counts['delayed']);
    }

    public function testJobWithAttemptsOption(): void
    {
        $job = $this->queue->add('retry-job', ['foo' => 'bar'], [
            'attempts' => 5,
        ]);

        $this->assertEquals(5, $job->opts->attempts);
    }

    public function testJobWithBackoffOption(): void
    {
        $job = $this->queue->add('backoff-job', ['foo' => 'bar'], [
            'attempts' => 3,
            'backoff' => [
                'type' => 'exponential',
                'delay' => 1000,
            ],
        ]);

        $this->assertEquals(3, $job->opts->attempts);
        $this->assertEquals('exponential', $job->opts->backoff['type']);
        $this->assertEquals(1000, $job->opts->backoff['delay']);
    }

    public function testJobWithRemoveOnCompleteOption(): void
    {
        $job = $this->queue->add('removable-job', ['foo' => 'bar'], [
            'removeOnComplete' => true,
        ]);

        $this->assertTrue($job->opts->removeOnComplete);

        // Test with count
        $job2 = $this->queue->add('removable-job-2', ['foo' => 'bar'], [
            'removeOnComplete' => 100,
        ]);

        $this->assertEquals(100, $job2->opts->removeOnComplete);
    }

    public function testCanUseCustomPrefix(): void
    {
        $customQueue = new Queue('custom-prefix-queue', [
            'prefix' => 'myapp',
            'connection' => [
                'host' => getenv('REDIS_HOST') ?: 'localhost',
                'port' => (int)(getenv('REDIS_PORT') ?: 6379),
            ],
        ]);

        $job = $customQueue->add('test-job', ['foo' => 'bar']);
        $this->assertNotNull($job->id);

        // Verify qualified name uses custom prefix
        $qualifiedName = $customQueue->getQualifiedName();
        $this->assertStringStartsWith('myapp:', $qualifiedName);

        $customQueue->obliterate(['force' => true]);
        $customQueue->close();
    }

    public function testCanShareConnection(): void
    {
        $connection = new RedisConnection([
            'host' => getenv('REDIS_HOST') ?: 'localhost',
            'port' => (int)(getenv('REDIS_PORT') ?: 6379),
        ]);

        $queue1 = new Queue('shared-conn-queue-1', ['connection' => $connection]);
        $queue2 = new Queue('shared-conn-queue-2', ['connection' => $connection]);

        $job1 = $queue1->add('job-1', ['queue' => 1]);
        $job2 = $queue2->add('job-2', ['queue' => 2]);

        $this->assertNotNull($job1->id);
        $this->assertNotNull($job2->id);

        $queue1->obliterate(['force' => true]);
        $queue2->obliterate(['force' => true]);

        // Queues don't own the connection, so they shouldn't close it
        $queue1->close();
        $queue2->close();

        // Connection should still be usable
        $result = $connection->getClient()->ping();
        $this->assertNotNull($result);

        $connection->close();
    }

    public function testGetJobsByMultipleTypes(): void
    {
        $this->queue->add('waiting-job', ['type' => 'waiting']);
        $this->queue->add('delayed-job', ['type' => 'delayed'], ['delay' => 60000]);

        // Get jobs from multiple types at once
        $jobs = $this->queue->getJobsByType(['waiting', 'delayed']);

        $this->assertGreaterThanOrEqual(2, count($jobs));
    }

    public function testJobDataIntegrity(): void
    {
        $complexData = [
            'string' => 'hello world',
            'number' => 42,
            'float' => 3.14159,
            'boolean' => true,
            'null' => null,
            'array' => [1, 2, 3],
            'nested' => [
                'foo' => 'bar',
                'baz' => [1, 2, 3],
            ],
            'unicode' => '你好世界 🌍',
        ];

        $job = $this->queue->add('complex-data-job', $complexData);
        $retrieved = $this->queue->getJob($job->id);

        $this->assertEquals($complexData['string'], $retrieved->data['string']);
        $this->assertEquals($complexData['number'], $retrieved->data['number']);
        $this->assertEquals($complexData['float'], $retrieved->data['float']);
        $this->assertEquals($complexData['boolean'], $retrieved->data['boolean']);
        $this->assertNull($retrieved->data['null']);
        $this->assertEquals($complexData['array'], $retrieved->data['array']);
        $this->assertEquals($complexData['nested'], $retrieved->data['nested']);
        $this->assertEquals($complexData['unicode'], $retrieved->data['unicode']);
    }
}
